src/parquet/properties.h

```diff
@@ -38,6 +39,101 @@ struct ParquetVersion {
 static int64_t DEFAULT_BUFFER_SIZE = 0;
 static bool DEFAULT_USE_BUFFERED_STREAM = false;
 
+class PARQUET_EXPORT ColumnEncryptionProperties {
+ public:
+  ColumnEncryptionProperties() = default;
+  ColumnEncryptionProperties(bool encrypt, std::string path)
+      : encrypt_(encrypt), path_(path), encrypted_with_footer_key_(encrypt) {}
+
+  bool encrypted() { return encrypt_; }
+  bool encrypted_with_footer_key() { return encrypted_with_footer_key_; }
+  std::string key() { return key_; }
+  std::string key_metadata() { return key_metadata_; }
+
+  void set_encryption_key(std::string key, uint32_t key_id = 0) {
+    std::string key_metadata =
+        key_id == 0 ? "" : std::string(reinterpret_cast<char*>(&key_id), 4);
+    set_encryption_key(key, key_metadata);
+  }
+
+  void set_encryption_key(std::string key, std::string key_metadata) {
+    if (!encrypt_) throw ParquetException("Setting key on unencrypted column: " + path_);
+    if (key.empty()) throw ParquetException("Null key for " + path_);
+
+    encrypted_with_footer_key_ = false;
+    key_ = key;
+    key_metadata_ = key_metadata;
+  }
+
+  std::string path() { return path_; }
+
+ private:
+  bool encrypt_;
+  bool encrypted_with_footer_key_;
+  std::string key_;
+  std::string key_metadata_;
+  std::string path_;
+};
+
+class PARQUET_EXPORT FileDecryptionProperties {
+ public:
+  FileDecryptionProperties(std::string footer_key) : footer_key_(footer_key) {
+    if (footer_key_.empty()) throw ParquetException("Decryption: null footer key");
+    if (!(footer_key_.length() == 16 || footer_key_.length() == 24 ||
+          footer_key_.length() == 32)) {
+      throw ParquetException("Wrong key length " + footer_key_.length());
+    }
+  }
+
+  FileDecryptionProperties(std::shared_ptr<DecryptionKeyRetriever> key_retriever)
+      : key_retriever_(key_retriever) {}
+
+  void set_aad(std::string aad) { aad_ = aad; }
+
+  void set_column_key(std::string name, std::string key) {
+    set_column_key(std::vector<std::string>({name}), key);
+  }
+
+  void set_column_key(std::vector<std::string> paths, std::string key) {
+    if (key.empty()) throw ParquetException("Decryption: null column key");
+    if (key.length() != 16 && key.length() != 24 && key.length() != 32)
+      throw ParquetException("Wrong key length " + key.length());
+
+    for (auto path = paths.begin(); path != paths.end(); path++) {
+      column_keys_[*path] = key;
+    }
+  }
+
+  std::string column_key(std::string path, std::string key_metadata = "") {
+    if (key_metadata.empty()) {
+      return column_keys_[path];
+    }
+    if (key_retriever_ == nullptr) {
+      throw ParquetException("no key retriever is provided for column key metadata");
```

Comment:
Could this happen (without the developer doing something wrong)?

Reply:
If developer did nothing wrong, this should not happen. I'll use DCHECK instead.